tests/kubesonic/test_k8s_join_disjoin.py

```diff
@@ -0,0 +1,387 @@
+
+def mark_minikube_started(vmhost):
+    logger.info("Mark minikube as started")
+    vmhost.shell("mkdir -p /run/minikube && touch /run/minikube/started")
+    logger.info("Minikube is marked as started")
+
+
+def mark_minikube_completed(vmhost):
+    logger.info("Mark minikube as completed")
+    vmhost.shell("rm -f /run/minikube/started")
+    logger.info("Minikube is marked as completed")
+
+
+def check_minikube_setup_started(vmhost):
+    logger.info("Check if minikube setup is started")
+    # Check the file createion time
+    minikube_setup_started = vmhost.shell("stat -c %z /run/minikube/started", module_ignore_errors=True)
+    if minikube_setup_started["stdout"] == "":
+        logger.info("Minikube setup is not started")
+        return False, None
+    else:
+        time_format = '%Y-%m-%d %H:%M:%S'
+        creation_time = datetime.strptime(minikube_setup_started["stdout"].strip().split('.')[0], time_format)
+        logger.info(f"Minikube setup is started, creation time is {creation_time}")
+        return True, creation_time
+
+
+def clean_up_and_setup_minikube(vmhost, creds):
+    # Mark the minikube as started
+    mark_minikube_started(vmhost)
+    # Clean up the previous minikube
+    remove_minikube(vmhost)
+    # Download minikube
+    download_minikube(vmhost, creds)
+    # Clean up the setup env
+    remove_k8s_master(vmhost)
+    # Setup k8s master
+    setup_k8s_master(vmhost, creds)
+    # Mark the minikube as completed
+    mark_minikube_completed(vmhost)
+
+
+def clean_configdb_k8s_table(duthost):
+    logger.info("Start to clean k8s table in configdb")
+    duthost.shell("sonic-db-cli CONFIG_DB DEL 'KUBERNETES_MASTER|SERVER'")
+    logger.info("K8s table in configdb is cleaned")
+
+
+@pytest.fixture()
+def setup_and_teardown(duthost, vmhost, creds):
+    check_dut_k8s_version_supported(duthost)
+    logger.info("Start to setup test environment")
+
+    # Get duthost asic type
+    asic_type = duthost.facts["asic_type"]
+    logger.info(f"DUT ASIC type is {asic_type}")
+
+    if asic_type == "vs":
+        clean_up_and_setup_minikube(vmhost, creds)
+    else:
+        while not check_minikube_ready(vmhost):
+            setup_started, creation_time = check_minikube_setup_started(vmhost)
+            if not setup_started:
+                clean_up_and_setup_minikube(vmhost, creds)
+                break
+            else:
+                # Generally, the setup process should be completed within 2 minutes
+                time_now = datetime.now()
+                logger.info(f"Current time is {time_now}")
+                time_diff = time_now - creation_time
+                time_diff_seconds = time_diff.total_seconds()
+                logger.info(f"Minikube setup has started for {time_diff_seconds} seconds")
+                if time_diff_seconds > MINIKUBE_SETUP_MAX_SECOND:
+                    logger.info("Minikube setup timeout, need to re-setup")
+                    clean_up_and_setup_minikube(vmhost, creds)
```

Comment:
IIUC, this while loop is for the file lock described in the readme. Here, if one testbed which first requested minikube master has taken more time than expected, then we restart the minikube master setup triggered by a second testbed. Is that right? 

Depending on the reason for the delay, isn't it possible to still run into the minikube master setup conflict here if the first operation was not cancelled? I think we should fail for testbed1 if testbed1 requests and minikube takes longer than expected to start. Without this failure, we may also run into an infinite loop

Reply:
Yes, you are right. The download step and setup step both have timeout(6 mins) now, it will fail if it's too long.